internal/job/executor.go

```diff
@@ -1032,8 +1007,7 @@ func (e *Executor) defaultCommandPhase(ctx context.Context) error {
 		e.shell.Promptf("%s", cmdToExec)
 	}
 
-	err = e.shell.RunWithoutPrompt(ctx, cmd[0], cmd[1:]...)
-	return err
+	return e.shell.RunWithoutPrompt(ctx, cmd[0], cmd[1:]...)
```

Comment:
This change breaks the `defer` near the start of the function (introduced in 3eaf791):

```go
	var err error
	defer func() { span.FinishWithError(err) }()
```

If we want to do this cleanup, we'd need to make `err` a named return instead.

Reply:
I think the same thing will happen here: https://github.com/buildkite/agent/blob/745645930b7d322eb9fc3660a0bb52b9edc945cd/internal/job/executor.go#L683 and possibly several other places yeah?

I think using named return values is an option to fix it: https://goplay.tools/snippet/yHRT8QgN8BG. Then this code can stay as is.

Reply:
And I think this is the relevant part of the spec:

> A "return" statement that specifies results sets the result parameters before any deferred functions are executed.

which appears under: https://go.dev/ref/spec#ReturnStmt

Reply:
Yep, I think we're on the same page there.